scio-avro/src/main/scala/com/spotify/scio/avro/types/SchemaUtil.scala

```diff
@@ -45,40 +43,52 @@ object SchemaUtil {
       case STRING | ENUM => ("String", Seq.empty)
       case BYTES => ("ByteString", Seq.empty)
       case ARRAY =>
-        val (fieldType, nested) = getFieldType(fieldName, fieldSchema.getElementType, indent)
+        val (fieldType, nested) =
+          getFieldType(className, fieldName, fieldSchema.getElementType, indent)
         (s"List[$fieldType]", nested)
       case MAP =>
-        val (fieldType, nested) = getFieldType(fieldName, fieldSchema.getValueType, indent)
-        (s"Map[String, $fieldType]", nested)
+        val (fieldType, nested) =
+          getFieldType(className, fieldName, fieldSchema.getValueType, indent)
+        (s"Map[String,$fieldType]", nested)
       case UNION =>
         val unionTypes = fieldSchema.getTypes.asScala.map(_.getType).distinct
         if (unionTypes.size != 2 || !unionTypes.contains(NULL)) {
           throw new IllegalArgumentException(
             s"type: ${fieldSchema.getType} is not supported. " +
             s"Union type needs to contain exactly one 'null' type and one non null type.")
         }
-        val (fieldType, nested) = getFieldType(fieldName, fieldSchema.getTypes.get(1), indent)
+        val (fieldType, nested) =
+          getFieldType(className,
+            fieldName,
+            fieldSchema.getTypes.asScala.filter(_.getType != NULL).head,
+            indent)
         (s"Option[$fieldType] = None", nested)
       case RECORD =>
-        val className = NameProvider.getUniqueName(fieldSchema.getName)
-        val nested = getCaseClass(fieldSchema.getFields, className, indent)
-        (className, Seq(nested))
+        val nestedClassName = s"$className$$${fieldSchema.getName}"
+        val nested =
+          getCaseClass(nestedClassName,
+            fieldSchema,
+            indent)
+        (nestedClassName, Seq(nested))
       case t => throw new IllegalArgumentException(s"Type: $t not supported")
     }
   }
   // scalastyle:on cyclomatic.complexity
 
-  private def getCaseClass(fields: JList[Field], name: String, indent: Int): String = {
-    val xs = fields.asScala
+  private def getCaseClass(className: String,
```

Comment:
You can get class name with `schema.getName()`?
Looks like `getCaseClass` is being used recursively, make an entry point version without `className`?

Reply:
I can't do that because I need to pass nested class name (ex. `ParentClass$NestedClass`) to getCaseClass function.